### PR TITLE
Share the doublings when the scalars are few, and count them

### LINKED ISSUE #212: Add Strauss-wNAF next to Bos-Coster in _multi_mult, and measure Pippenger above a threshold
`_multi_mult` is Bos-Coster (`curves/curve_group.py:730`, sourced from
`https://cr.yp.to/badbatch/boscoster2.py`). libsecp256k1, which is the authority
on every other secp256k1 answer in this library, does not use it: it implements
**Strauss-wNAF** and **Pippenger**, and dispatches between them in
`secp256k1_ecmult_multi_var` (`src/ecmult_impl.h:823`) with a three-level
fallback — Pippenger when the scratch space allows and the batch has ≥ 88
points, else Strauss, else `ecmult_multi_simple_var`, the naive loop. The
threshold is a measured constant and says so:

```c
/* Minimum number of points for which pippenger_wnaf is faster than strauss wnaf */
#define ECMULT_PIPPENGER_THRESHOLD 88
```

Their Strauss is GLV as well: `secp256k1_scalar_split_lambda` splits each scalar
into `na_1 + na_lam·λ` of ~128 bits, `WINDOW_A 5`, wNAF length fixed at
`CEIL_DIV(256, w)`.

## Why this is worth doing

The cost of both of their algorithms depends only on the bit length and the
window size — a fixed number of wNAF digits, a fixed number of buckets — so **no
degenerate input exists**. Bos-Coster's does depend on the values: its step is
subtractive, and two scalars of distant magnitude cost `n₁/n₂` iterations. That
is #175, where `multi_mult([10**6, 1], [G, H])` is measured at 10.58 s and
`multi_mult([-1, 1], [G, H])` does not finish at all. Replacing the algorithm
makes that class of defect impossible rather than fixed.

This cannot be delegated to the bindings, which is what makes it our problem:
`secp256k1_ecmult_multi_var` is internal, `include/` exposes no multi-scalar
multiplication, and the only public building blocks are
`secp256k1_ec_pubkey_tweak_mul` and `secp256k1_ec_pubkey_combine` — k full
scalar multiplications plus an addition, which is precisely what a multi-scalar
algorithm exists to avoid. See #128 for what *is* delegable.

## What to do

- Strauss-wNAF first: it is the branch that covers btclib's own use. With a
  threshold of 88 points, batch verification — 2k scalars for k signatures —
  stays in the Strauss region up to 44 signatures.
- Pippenger second, and only if a benchmark on this codebase asks for it. Do not
  copy 88: their threshold is measured on C with dedicated field arithmetic, and
  ours would be measured on python integers. The number has to be ours or it is
  decoration.
- The GLV λ-split is a separate decision, not a prerequisite: `curve_group_2.py`
  already collects the reading on window methods and joint sparse form, and #171
  is the adjacent work on point addition.
- Benchmark against the current implementation at k = 2, 3, 5, 10 and on
  `ssa.assert_batch_as_valid_`, and record the numbers in the commit message.
  `_multi_mult` has one internal caller (`ecc/ssa.py:449`) and one public entry
  point (`curves.multi_mult`), so the surface to keep is small.

## Relation to #175

#175 is the hang, and it has a five-line cure: an Euclidean step
(`divmod` instead of subtraction) that terminates in `O(log n)`. If that lands
first, this issue deletes it; if this lands first, #175 closes with it. Either
order is defensible — what is recorded in #175 is that the two are alternatives
rather than a sequence, and that the choice is about how long the hang stays
reachable.

### PR DESCRIPTION
**Pull request by fametrano**
`_multi_mult` was Bos-Coster whatever the batch. It is now Strauss'
interleaved wNAF below fifty-six scalars and Bos-Coster from there on:
one width-w NAF and one table of odd multiples per point, and a single
left-to-right loop sharing one doubling per bit position among all of
them. Both implementations stay in the library — `_multi_mult_w_NAF` and
`_multi_mult_bos_coster` — as the issue asks: btclib is didactic as much
as it is a library, and the heap one is the twenty lines that can be
read. `_multi_mult` is the dispatch, and it is the only thing that had
to be decided.

No GLV split (curve-specific, and this path serves every curve; the
doublings it would halve are already shared) and no Pippenger, on the
measurement below.

## What it costs and what it buys

Best of interleaved reps, random 256-bit scalars on secp256k1, python
3.14, macOS arm64, ms:

| scalars | Bos-Coster | wNAF w=5 | naive | speedup |
| --- | --- | --- | --- | --- |
| 2 | 3.91 | 1.99 | 3.11 | 1.96x |
| 3 | 4.56 | 2.59 | 4.58 | 1.76x |
| 5 | 6.34 | 3.92 | 7.74 | 1.62x |
| 10 | 10.02 | 6.96 | 15.62 | 1.44x |
| 32 | 23.36 | 20.45 | 49.75 | 1.14x |
| 56 | 36.13 | 36.30 | 90.09 | 1.00x |
| 64 | 41.71 | 41.99 | 100.65 | 0.99x |
| 128 | 70.69 | 80.83 | 202.86 | 0.87x |

`ssa.assert_batch_as_valid_`, which spends two scalars per signature:

| signatures | scalars | Bos-Coster | dispatched | speedup |
| --- | --- | --- | --- | --- |
| 2 | 4 | 7.79 | 5.93 | 1.31x |
| 3 | 6 | 9.65 | 7.32 | 1.28x |
| 5 | 10 | 12.77 | 10.01 | 1.28x |
| 10 | 20 | 20.38 | 16.87 | 1.21x |
| 16 | 32 | 28.28 | 24.74 | 1.14x |
| 32 | 64 | 47.75 | 47.60 | 1.00x |
| 64 | 128 | 82.00 | 82.14 | 1.00x |

The threshold is where the two meet, measured and not borrowed:
wNAF over Bos-Coster is 0.91 at 40 scalars, 0.94 at 48, 1.000 at 56,
1.001 at 64, 1.04 at 72, 1.09 at 96. Point operations say the same in a
form no machine load can move: 44.5 per scalar for Bos-Coster over 128
scalars against 53.3 for the wNAF, and 270 against 178 over two. `w=5`
by the same measurement — 57.7 operations per scalar at w=4, 53.3 at
w=5, 55.1 at w=6; over ten scalars 7.81 ms, 7.27 and 7.38.

The issue's bar was 0.54 ms per scalar above 32 scalars, and this does
not beat it: the wNAF's own per-scalar cost is ~0.64 ms and what it
saves is the fixed one, so the win is on few scalars and the dispatch is
the honest way to keep it.

## Pippenger: measured, and not implemented

Prototyped with sparse buckets, the running-sum trick over the nonempty
ones, and the best window per size. It lost at every size tried, against
Bos-Coster: 162 ms vs 125 at 256 scalars, 284 vs 225 at 512, 500 vs 406
at 1024 (best windows c=6, c=7, c=8). On python integers a bucket sum is
an addition of points that costs what any addition costs, where in C it
is the cheap part, so its crossover — if it has one — is past any batch
btclib is asked to verify. Copying libsecp256k1's 88 would have been
decoration. The numbers and the reasoning are in `_multi_mult`'s
docstring, so the next person does not have to run it again.

## Issue #175's cure stays

The whole of Euclid in the Bos-Coster step is what keeps that path from
hanging on a distant pair of magnitudes, and that path is live above the
threshold: deleting it would put the hang back for exactly the batches
it now answers. The interleaved wNAF is indifferent to magnitude in the
first place, its cost being a function of bit length alone.
`multi_mult([10**6, 1])`, `([-1, 1])` — which is `[n-1, 1]` after the
public wrapper reduces mod n, the contract being that negatives are
reduced there and refused below it — and `([n-1, 1])` are asserted
against both implementations, and run in milliseconds.

## Verification

- `uv run pytest` — exit 0, 14791 passed
- `uv run pre-commit run --all-files` — exit 0
- `uv run --locked --no-default-groups --group docs sphinx-build -W
  --keep-going -b html docs/source docs/build/html` — exit 0 on a clean
  `docs/build`, which is where it first failed: `(abs(d) - 1)` in a
  docstring, because docutils reads the pipes of the mathematical
  spelling as a substitution reference
- coverage of `btclib/curves/curve_group.py` is 100.00%: the tests cover
  both table branches (w of 1 and 2 against 3 and 5), the zero scalar,
  the point at infinity, the negative digits, a wNAF shorter than the
  longest, and both sides of the threshold
- the two implementations are checked against each other and against the
  plain sum of scalar multiplications, exhaustively over the scalar pairs
  of two low-cardinality curves and with fixed-seed random scalars on
  every curve of the catalogue, not only secp256k1

The wNAF recoding and the table of odd multiples move to `curve_group`,
next to `convert_number_to_base` and for the same reason: the
interleaved multi multiplication needs them and `curve_group` cannot
import `curve_group_2`. `double_mult_w_NAF` builds its two tables with
that same function now.

Sibling PRs move the CHANGELOG entry count too; re-derive it at merge
with `grep -c '^- ' CHANGELOG.md`.

Closes #212.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by Sourcery

Dispatch multi scalar multiplication between interleaved wNAF and Bos-Coster based on batch size to improve performance for small scalar sets while preserving the existing Euclid-based fix for pathological magnitudes.

New Features:
- Add an interleaved width-w wNAF-based multi scalar multiplication implementation that shares doublings across points.
- Expose a configurable Bos-Coster threshold constant to switch algorithms by scalar count.

Enhancements:
- Refactor multi-mult argument validation and zero-scalar handling into a shared helper used by both implementations.
- Move wNAF recoding and odd-multiples table construction into the core curve_group module for reuse across implementations.
- Reuse the shared odd-multiples helper in double_mult_w_NAF to simplify table construction and align behavior.

Documentation:
- Update CHANGELOG and HISTORY to document the new multi-mult dispatch behavior and its measured performance characteristics.
- Clarify curve.multi_mult docstring to describe algorithm selection between interleaved wNAF and Bos-Coster.

Tests:
- Add comprehensive tests verifying interleaved wNAF multi multiplication against Bos-Coster and the plain sum-of-mults across curves, widths, and edge cases.
- Add dispatch tests to ensure _multi_mult chooses the correct algorithm around the threshold and shares identical error semantics.
- Add regression tests covering distant scalar magnitudes to ensure the Bos-Coster path remains bounded and matches expected results.

### CODE REVIEW

File: btclib/curves/curve_group.py [hunk 1]
```diff
@@ -547,6 +547,69 @@ def convert_number_to_base(i: int, base: int) -> list[int]:
     return digits[::-1]
 
 
+def mods(m: int, w: int) -> int:
+    """Signed modulo function."""
+    w2: int = pow(2, w)
+    M = m % w2
+    return M - w2 if M >= (w2 / 2) else M
```
[the remark this answers is not in the document]
Comment (fametrano) [reply btclib/curves/curve_group.py]: Done, and it had already been done in the other copy: `dev`'s ceca1299 fixed exactly this line in `curve_group_2.py` while this branch was moving the function to `curve_group.py`, so the rebase would have reverted it. `mods` now reads `M >= (w2 // 2)` in its new home.

Floor division rather than the shift, though, and `pow(2, w)` kept: `//` is what the rest of the module writes (`b // 4`, `b // 2`, `(M[j] - 1) // 2`), and the point of the change is the arithmetic, not the spelling of the halving.

The float was not a wrong answer today — `pow(2, w)` is exactly representable and CPython compares int against float exactly — but the conversion is a limit the arithmetic does not have: past `w = 1024` it raises `OverflowError` where the domain error is `w` itself.

File: tests/curves/curve_group_test.py [hunk 2]
```diff
@@ -405,6 +410,127 @@ def test_assorted_jac_mult() -> None:
...
+    Exhaustive over the scalar pairs of two low-cardinality curves, for
+    the widths where the table of odd multiples has one entry (w of 1 and
+    2, digits of +-1 only) and where it has more, and with the zero
+    scalar, the point at infinity and a scalar of one digit against one
+    of five -- the case where a wNAF runs out of digits before the loop
+    does.
+    """
+    for ec in (low_card_curves["ec13_11"], ec23_31):
+        HJ = jac_from_aff(second_generator(ec))
+        for w in (1, 2, 3, 5):
```
[the remark this answers is not in the document]
Comment (fametrano) [reply tests/curves/curve_group_test.py:472]: Not as written: `MAX_W = 5` is the size cap of `cached_multiples`' table, used by `mult_fixed_window_cached`. It has nothing to do with `_multi_mult_w_NAF`'s `w`, which has no maximum at all — only `w <= 0` raises. The tuple's `5` is already the default width, and `MAX_W` would just be the same number arriving under a misleading name.

The point underneath it is a good one, so the extreme it should reach is a different one: a window **wider than the curve's own order**, where the table of odd multiples wraps past `n` and holds the point at infinity among its entries — 3 of 32 for `n = 11` at `w = 7`, 2 of 64 for `n = 31` at `w = 8`. That is now tested, against `_sum_of_mults` on both low-cardinality curves.

It sits outside the exhaustive `k1 × k2` loop rather than in the width tuple, which is the runtime concern you raised: the table is `2^(w-2)` entries and every one of the 961 pairs would pay to build it, for digits the `w = 5` pass already covers.

File: CHANGELOG.md [hunk 3]
```diff
@@ -1706,6 +1706,33 @@ twenty-nine source-breaking changes on their own.
...
+  at, so 56 is where the two measure the same and the dispatch changes
+  hands. Both implementations stay, tested against each other and against
+  the plain sum of scalar multiplications on every curve of the
+  catalogue; the threshold is a measurement, not libsecp256k1's 88, whose
+  algorithms and machine are not these. Pippenger is *not* in the tree
+  for the same reason: prototyped with sparse buckets and the running-sum
+  trick, best window per size, it lost at every size measured — 162 ms
+  against Bos-Coster's 125 at 256 scalars, 284 against 225 at 512, 500
+  against 406 at 1024 — because in python its bucket sums are additions
+  like any other, where in C they are the cheap part (issue #212)
```
[the remark this answers is not in the document]
Comment (fametrano) [reply CHANGELOG.md:1924]: Declining this one: it would be the only capital in the file. `CHANGELOG.md` writes `python` in lower case 38 times and `Python` zero times — `grep -c '\bpython\b' CHANGELOG.md` against `grep -c '\bPython\b'` — so the correction would make this entry the inconsistent one rather than the other thirty-seven.

**Comment by fametrano** [comment]
Thanks — both of the high-level points landed, one of them further than the review put it. Rebased onto `dev` and force-pushed; the three inline comments have replies of their own.

**`mods` and the float.** Fixed, and it needed the rebase to survive: `dev`'s ceca1299 had already made this change in `curve_group_2.py` while this branch was moving the function to `curve_group.py`, so the merge would otherwise have carried the float back in. `w2 // 2` rather than `w2 >> 1`, `//` being what the rest of the module writes.

**Dispatching on the nonzero pairs.** Taken, and the docstring that argued against it was wrong. It claimed a caller passing zeros "pays 1% for it"; measured, on secp256k1, best of five interleaved reps:

| scalars passed | nonzero | Bos-Coster | interleaved wNAF | cost of dispatching on `len()` |
| ---: | ---: | ---: | ---: | ---: |
| 56 | 2 | 1.81 ms | 1.02 ms | 1.78x |
| 56 | 4 | 2.30 ms | 1.48 ms | 1.55x |
| 128 | 8 | 3.27 ms | 2.42 ms | 1.35x |
| 1000 | 2 | 1.85 ms | 0.98 ms | 1.89x |

The zeros are dropped in `_multi_mult_pairs` before either implementation sees them, so the batch that does the work is the nonzero one and it is the one the threshold is about: 56 scalars of which 2 are nonzero is a batch of two. `_multi_mult` now counts them out, `sum(1 for n in scalars if n) < BOS_COSTER_THRESHOLD`, which costs 0.8 us at 56 scalars and 12.8 us at 1000 — under a thousandth of either algorithm's time at those sizes.

One consequence worth naming: it takes Bos-Coster's all-zero early return away from its only caller, a batch of nothing but zeros being a batch of no nonzero scalars and so below any threshold. The test now asks both implementations for that case by name; without it the coverage of `curve_group.py` falls to 99.76%.

The dispatch test also draws its scalars from 1 rather than 0 now. On `ec23_31` a zero among the 56 was likely, and it would have moved the batch to the other side of the threshold with the test still passing — exercising one branch twice and reporting it as both.

Gates, all exit 0: `uv run pytest` (14932 passed), `uv run pre-commit run --all-files`, and the sphinx `-W` docs build on a clean `docs/build`. Coverage of `curve_group.py` is back to 100.00%.
